### PR TITLE
feat: Determine planned actions

### LINKED ISSUE #2340: Implement planned action to play music
**Describe the feature you'd like**
Implement the feature to play music, after a scheduler. Like suppose a query is made ```set alarm in one minute to play music```, the response received is of the format ```alarm set in one minute to play music```. The client should be able to handle this response. The app should be able to start a scheduler of 1 minute, after completion of which the susi app should execute the task to play the music.

**Additional context**
This is in reference to #2336 and as per discussion in the meetings.

**Would you like to work on the issue?**
Yes

Checklists:
- [ ] Determine if its a planned action or not
- [ ] Implement scheduler in background
- [ ] Execute the task in the scheduler

### PR DESCRIPTION
**Pull request by atm1504**
Fixes #2340 

Changes: 
- [x] Determine if its a planned action or not
- [x] Implement scheduler in background
- [x] Execute the task in the scheduler
- [x] Determine a different kind of plan actions
- [x] Show data accordingly if planned action or not
- [x] Extend the plan actions to perform other tasks

Screenshots for the change: 
https://photos.app.goo.gl/TiNCGXmjpdGr3Mey7
Plan action

Apk:
[app-fdroid-debug.apk.zip](https://github.com/fossasia/susi_android/files/3489122/app-fdroid-debug.apk.zip)

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt [hunk 1]
```diff
@@ -529,6 +534,24 @@ class ChatPresenter(context: Context) :
         computeOtherMessage()
     }
 
+    override fun determinePlanAction(response: SusiResponse?, parseSusiHelper: ParseSusiResponseHelper) {
+        val query = response?.query.toString()
+        identifier = parseSusiHelper.identifier
+        if (query.contains(ALARM, false) || query.contains(PLAN, false)) {
+            // Perform the task thats need to be done after one minute
+            delayIdentifier = identifier
+            planHandler.postDelayed(delayRunnable, 60000) // Time value will change
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: Why use handler?

Comment (atm1504) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: I wanted to communicate with the main thread by setting a scheduler. But at the same time, I didn't want the other processes to stop. The user should be able to use the app normally. The scheduler would run in the back ground, and perform the task after the time is over. So I used handler

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: I don't know who has spread this lie that handler tasks tun in background. They don't

They are run on main thread

Comment (atm1504) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: If that is the case, I will surely re factor the code. And write it in background thread seperately.

Comment (atm1504) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: If I am not wrong, handlers provide the way to interact with the main UI thread. Here in this feature our main task is to perform a task, that will interact with ui changes. Like for ```plan to tell a joke in 1 minute``` the scheduler should display a joke in the char after 1 minute. So won't it be good to use handler here?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: Yes, it will be. But here, you're just statically writing `postDelayed(..., 60000)`

Comment (atm1504) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: The present skills are framed for testing purpose and the scheduler is fixed for 1 minute. No, methods have been implemented in the server to detect the time and send that response. So, I have statically set the delayed to 1 minute. Once, that is implemented in the server, I will change that time accordingly in the client-side.

**Comment by atm1504** [comment]
@iamareebjamal please review it. For ```set alarm in one minute``` query the response of delay comes from the server. In the video I have used 5sec, for testing purpose only.

**Comment by atm1504** [comment]
@iamareebjamal please review it

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt [hunk 2]
```diff
@@ -457,59 +469,85 @@ class ChatPresenter(context: Context) :
...
+                                        mapData = parseSusiHelper.mapData,
+                                        isHavingLink = parseSusiHelper.isHavingLink,
+                                        datumList = parseSusiHelper.datumList,
+                                        webSearch = parseSusiHelper.webSearch,
+                                        tableItem = tableItem,
+                                        identifier = identifier,
+                                        skillLocation = susiResponse.answers[0].skills[0]
+                                ), this)
+                            } catch (e: Exception) {
+                                Timber.e(e)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: Pass in a useful message as context

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt [hunk 3]
```diff
@@ -457,59 +469,85 @@ class ChatPresenter(context: Context) :
...
-                        var speechReply = setMessage
-                        if (parseSusiHelper.isHavingLink) {
-                            speechReply = setMessage.substring(0, setMessage.indexOf("http"))
+            for (j in 0 until actionSize) {
+                val parseSusiHelper = ParseSusiResponseHelper()
+                parseSusiHelper.parseSusiResponse(susiResponse, j, utilModel.getString(R.string.error_occurred_try_again))
+
+                if (parseSusiHelper.actionType == Constant.VIDEOPLAY) {
+                    isVideo = true
+                    Log.d("KHANKI", "Is Video")
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: Remove

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt [hunk 4]
```diff
@@ -457,59 +469,85 @@ class ChatPresenter(context: Context) :
...
-                        identifier = parseSusiHelper.identifier
-                        chatView?.playVideo(identifier)
-                    } else if (parseSusiHelper.actionType == Constant.ANSWER && (PrefManager.checkSpeechOutputPref() && check || PrefManager.checkSpeechAlwaysPref())) {
-                        setMessage = parseSusiHelper.answer
-
-                        var speechReply = setMessage
-                        if (parseSusiHelper.isHavingLink) {
-                            speechReply = setMessage.substring(0, setMessage.indexOf("http"))
+            for (j in 0 until actionSize) {
+                val parseSusiHelper = ParseSusiResponseHelper()
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt:472]: Why allocating in a loop?

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt [hunk 5]
```diff
@@ -529,6 +567,137 @@ class ChatPresenter(context: Context) :
...
+    }
+
+    override fun determineVideoPlanAction(susiResponse: SusiResponse, actionSize: Int) {
+        val handler = Handler()
+        handler.post { handleVideoAlarm(susiResponse, 0) }
+
+        planVideoSusiResponse = susiResponse
+        var j = 1
+        var plan_delay = 0L
+        while (j < actionSize) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt]: Use for loop

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt (hunk 4 above)
Comment (atm1504) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatPresenter.kt:472]: The actions part of the susi response contains a set of replies. Each of these replies needs to be parsed individually singr PareseSusiResponseHelper class. As it is necessary to identify the action type, whether it is a video or something else.

**Review by iamareebjamal — APPROVED** [review]

**Pull request merged**